Implement a rule to check deprecated hook implementations

Kingdutch opened the pull request:
This will analyse hook implementations in `.inc` and `.module` files. It detects a hook by looking at the pattern `mymodule_[hook]` which is detected from the file-name the function is in.

It will replace `mymodule` with `hook` and then try to find a function with that name. For this to work PHPStan Drupal's `DrupalAutoloader` class must be adjusted to load `.api.php` files as well so that the hooks can be discovered. In testing on the Open Social code-base this did not cause any issues.

The rule will provide the deprecation message in the error if it's provided, if no deprecation message was provided (i.e. a lonesome `@deprecated` tag) then the hook will simply be shown as deprecated without advice.

Fixes #126

mglaman commented on src/Rules/Deprecations/DeprecatedHookImplementation.php:
```diff
@@ -0,0 +1,68 @@
+    public function processNode(Node $node, Scope $scope) : array
+    {
+        assert($node instanceof Function_);
+        if (!str_ends_with($scope->getFile(), ".module") && !str_ends_with($scope->getFile(), ".inc")) {
```
phpstan-drupal still supports PHP 7.4 technically for Drupal 9. But we can trust the PHP 8 polyfill exists for D9, so it should be okay right?

bbrala replied:
Also, Drupal 9 officially doesn't support 7.4 anymore ;)

mglaman replied:
"officially"

bbrala replied:
Probably no breaking code, haven't noticed, just bo promised I think

Kingdutch replied:
I've used `str_ends_with` and `str_starts_with` for pretty much the entirety of Drupal 9's lifetime :D The `symfony/polyfill-php80` package takes care of the functions (and Drupal depends on it).

Not sure if we need to add a dependency on that package explicitly in phpstan-drupal?

mglaman replied:
Nah, I think we're OK and I should probably just drop PHP 7.4. I was waiting for PHPStan to drop it, but the Drupal community has pretty much moved onto PHP 8.1 (it seems.)

Nothing to do here, though.

mglaman reviewed (commented):
Awesome! This looks great

mglaman commented:
I am surprised this works given

```
        excludePaths:
                - '*.api.php'
```

mglaman commented:
@Kingdutch 

> For this to work PHPStan Drupal's DrupalAutoloader class must be adjusted to load .api.php files as well so that the hooks can be discovered. In testing on the Open Social code-base this did not cause any issues.

I don't see these changes? We'd have to add something here https://github.com/mglaman/phpstan-drupal/blob/main/src/Drupal/DrupalAutoloader.php#L110-L113 which is how we load post update files

Kingdutch commented:
Modified the DrupalAutoloader for `.api.php` discovery (I suspect I had that modified locally but it went missing because it wasn't in the `open_social` repo where I originally committed this change).

Moved the hook declarations to the fixture's `.api.php` file to prove that works :)

mglaman reviewed (approved):
This is amazing! I'll merge later today.
